pkg/controller/persistentvolume/state.go

```diff
@@ -0,0 +1,225 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package persistentvolume
+
+import (
+	"fmt"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/conversion"
+	"k8s.io/kubernetes/pkg/util/keymutex"
+)
+
+type persistentVolumeStateContext struct {
+	volume        *api.PersistentVolume
+	claim         *api.PersistentVolumeClaim
+	claimNotFound bool
+	client        binderClient
+	volumeIndex   *persistentVolumeOrderedIndex
+	locks         keymutex.KeyMutex
+	states        []persistentVolumeState
+}
+
+type persistentVolumeState interface {
+	isCurrentState(context *persistentVolumeStateContext) bool
+	processState(context *persistentVolumeStateContext) error
+}
+
+// newPersistentVolumeStateContext is a factory function that creates persistentVolumeStateContext objects with
+// the correct internal state object set according to the phase of the PersistentVolume.
+// This factor also attempts to find any claim bound to the volume via the volume's ClaimRef
+func newPersistentVolumeStateContext(pv *api.PersistentVolume, client binderClient, volumeIndex *persistentVolumeOrderedIndex, locks keymutex.KeyMutex) (*persistentVolumeStateContext, error) {
+	ctx := &persistentVolumeStateContext{
+		client:      client,
+		volume:      pv,
+		volumeIndex: volumeIndex,
+		locks:       locks,
+		states: []persistentVolumeState{
+			&persistentVolumePendingState{},
+			&persistentVolumeBoundState{},
+			&persistentVolumeReleasedState{},
+		},
+	}
+	if err := ctx.fetchClaim(); err != nil {
+		return nil, err
+	}
+	return ctx, nil
+}
+
+// fetchClaim will find the claim bound to the PV, if any, and set it on the persistentVolumeContext object.
+// any error that is not 404 will be returned.  Missing claims (404s) are normal and not fatal to the PV's state.
+// Claims that 404 are important to track, because a deleted claim is meaningful to a volume bound to it.
+func (ctx *persistentVolumeStateContext) fetchClaim() error {
+	if ctx.volume != nil && ctx.volume.Spec.ClaimRef != nil {
+		claim, err := ctx.client.GetPersistentVolumeClaim(ctx.volume.Spec.ClaimRef.Namespace, ctx.volume.Spec.ClaimRef.Name)
+		if err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+		if err != nil && errors.IsNotFound(err) {
+			ctx.claimNotFound = true
+			return nil
+		}
+		ctx.claim = claim
+		ctx.claimNotFound = false
+	}
+	return nil
+}
+
+//makeCloneForUpdate clones a volume that is then used to update the API.
+// a clone is used because modifying the PV itself means modifying the volumes held by the controller (it's a pointer)
+// as well as having to rollback those changes in the event of API failure.
+func makeCloneForUpdate(pv *api.PersistentVolume) (*api.PersistentVolume, error) {
```

Comment:
So, just to make sure I understand, the clone is simply a watcher on the API call that passes the return code from the API call back to the controller. If we aren't going to react to the return, why are we even tracking it? 

If the controller has to act on that return code, we then must rollback which requires us to understand the state and therefore order does become a concern here. Correct? I guess I mentally correlate this with database transactions. You can do a dirty read/write, you can do an insert-or-update-if doesn't exist but you do run the risk of synchronization issues which are pretty complex. Order does matter to these transactions if you are reading a row, updating the values and pushing your copy to the DB that overwrites a possible change in the process. I am wonder if we could see this happening as well with the rollback from a failed API call.